File: cocos2d/EscapeShrine/Classes/Interpreter.cpp
#include "Interpreter.h"
#include "StrUtil.h"

InterpreterNode::~InterpreterNode()
{
	L("InterpreterNode::destructor");
}

CommandNode::~CommandNode()
{
	L("CommandNode::destructor");
}

IfNode::~IfNode()
{
	L("IfNode::destructor");
	if ( ifCommand != nullptr )
	{
		for ( InterpreterNode* n : *ifCommand )
		{
			delete n;
		}
		delete ifCommand;
		ifCommand = nullptr;
	}

	if ( execCommand != nullptr )
	{
		for ( InterpreterNode* n : *execCommand )
		{
			delete n;
		}
		execCommand->clear();
		delete execCommand;
		execCommand = nullptr;
	}
	
	if ( elseNode != nullptr )
	{
		delete elseNode;
		elseNode = nullptr;
	}
}

Interpreter::~Interpreter()
{
	L("Interpreter::destructor");
	//event = nullptr;
}

CommandListNode* Interpreter::parse(std::string code)
{
	StringTokenizer st(code, " ");
	CommandListNode *ret = new CommandListNode();
	std::string tmp;
	while ( (tmp = st.next()) != "\0" ) {
		if ( tmp == "if" ) {
			IfNode* ifN = Interpreter::makeIf(st);
			ret->push_back(ifN);
		} else if ( tmp == "set") {
			CommandNode *n = new CommandNode(st.next(), st.next());
			ret->push_back(n);
		}
	}
	return ret;
}

void Interpreter::execEvent(CommandListNode *list)
{
	for ( InterpreterNode* n : *list )
	{
		if ( n->isIfNode() )
		{
			IfNode *ifNode = static_cast<IfNode*>(n);
			CommandListNode *execListNode = execIf(ifNode);
			if ( execListNode != nullptr )
			{
				for (unsigned int j = 0; j < execListNode->size(); j++) {
					CommandNode* cn = static_cast<CommandNode*>(execListNode->at(j));
					event->executeCommand(cn);
				}
			}
		}
		else
		{
			CommandNode *node = static_cast<CommandNode*>(n);
			event->executeCommand(node);
		}
	}
}


CommandListNode* Interpreter::execIf(IfNode *node)
{
	if ( node == nullptr ) {
		return nullptr;
	}
	bool flag = true;
	CommandListNode *cln = node->ifCommand;
	for (unsigned int i = 0; i < cln->size(); i++)
	{
		CommandNode* comm = static_cast<CommandNode*>(cln->at(i));
		flag &= event->executeIfCommand(comm);
	}
	if ( flag )
	{
		return node->execCommand;
	}
	else
	{
		return execIf(node->elseNode);
	}
}

IfNode* Interpreter::makeIf(StringTokenizer &st) {
	CommandListNode *ifCommand = new CommandListNode();
	CommandListNode *execCommand = new CommandListNode();
	std::string tmp;
	while ( (tmp = st.next()) != "\0" ) {
		if ( tmp == "isset" ) {
			CommandNode *cn = new CommandNode(st.next(), st.next());
			ifCommand->push_back(cn);
		} else if ( tmp == "set" ) {
			CommandNode *cn = new CommandNode(st.next(), st.next());
			execCommand->push_back(cn);
		} else if ( tmp == "else" ) {
			IfNode *ifNode = new IfNode();
			ifNode->elseNode = makeIf(st);
			ifNode->ifCommand = ifCommand;
			ifNode->execCommand = execCommand;
			return ifNode;
		} else if ( tmp == "end" ) {
			IfNode *ifNode = new IfNode();
			ifNode->ifCommand = ifCommand;
			ifNode->execCommand = execCommand;
			return ifNode;
		}
	}
	L("error");
    return nullptr;
}

File: java/WTK22/apps/RECT2ex/src/com/strnet/game/interpreter/Interpreter.java
/*
 * Last modified: 2009/06/10 13:56:18
 * author: Kazamai, Kou
 */
package com.strnet.game.interpreter;

import java.util.Vector;
import com.strnet.game.common.StringTokenizer;

/**
 * ’´ƒ~ƒjŒ¾Œê
 */
public class Interpreter
{
	public static CommandListNode parse(String code)
	{
		StringTokenizer st = new StringTokenizer(code, " \r\n\t");
		
		String tmp;
		CommandListNode list = new CommandListNode();
		while ( ((tmp = st.nextToken()) != null ) )
		{
			if ( tmp.equals("if") )
			{
				list.add(makeIf(st));
			}
			else if ( tmp.equals("set") )
			{
				list.add(makeSet(st));
			}
			else
			{
				throw new IllegalArgumentException(tmp);
			}
		}
		
		return list;
	}
	
	private static CommandNode makeSet(StringTokenizer st)
	{
		CommandNode node = new CommandNode();
		node.setCommand(st.nextToken());
		node.setArg(st.nextToken());
		return node;
	}
	
	private static IfNode makeIf(StringTokenizer st)
	{
		String tmp;
		IfNode ifNode = new IfNode();
		CommandListNode ifCommand = new CommandListNode();
		CommandListNode execCommand = new CommandListNode();
		
		while ( ((tmp = st.nextToken()) != null ) )
		{
			if ( tmp.equals("isset") )
			{
				ifCommand.add(makeSet(st));
			}
			else if ( tmp.equals("set") )
			{
				execCommand.add(makeSet(st));
			}
			else if ( tmp.equals("else") )
			{
				ifNode.setElseNode(makeIf(st));
				ifNode.setIfCommandListNode(ifCommand);
				ifNode.setExecCommandListNode(execCommand);
				return ifNode;
			}
			else if ( tmp.equals("end") )
			{
				ifNode.setIfCommandListNode(ifCommand);
				ifNode.setExecCommandListNode(execCommand);
				return ifNode;
			}
			else
			{
				throw new IllegalArgumentException(tmp);
			}
		}
		throw new IllegalArgumentException();
	}



	private InterpreterEvent event;
	
	public Interpreter(InterpreterEvent event)
	{
		this.event = event;
	}
	
	public void execEvent(CommandListNode list)
	{
		for ( int i = 0; i < list.getCommands().size(); i++ )
		{
			Node node = (Node)list.getCommands().elementAt(i);
			CommandListNode execListNode = null;
			if ( node instanceof IfNode )
			{
				execListNode = execIf((IfNode) node);
			}
			else if ( node instanceof CommandListNode )
			{
				execListNode = (CommandListNode) node;
			}
			else
			{
				CommandNode n = (CommandNode) node;
				event.executeCommand(n);
			}
		
			if ( execListNode != null )
			{
				for ( int j = 0; j < execListNode.getCommands().size(); j++ )
				{
					CommandNode n = (CommandNode) execListNode.getCommands().elementAt(j);
					event.executeCommand(n);
				}
			}
		}
	}
	
	public CommandListNode execIf(IfNode node)
	{
		if ( node == null )
		{
			return null;
		}

		boolean flag = true;
		if ( node.getIfCommandListNode() != null )
		{
			Vector v = node.getIfCommandListNode().getCommands();
			for ( int i = 0; i < v.size(); i++ )
			{
				CommandNode n = (CommandNode) v.elementAt(i);
				flag &= event.executeIfCommand(n);
			}
		}
		if ( flag )
		{
			return node.getExecCommandListNode();
		}
		else
		{
			return execIf(node.getElseNode());
		}
	}
}